vm_phys.c

```diff
@@ -0,0 +1,249 @@
+#include <vm_phys.h>
+#include <queue.h>
+#include <kmem.h>
+#include <libkern.h>
+
+
+#define ORDER_SIZE(x) (1 << (x))
+
+/* this is a stub, should be moved to other file */
+#define panic(MSG...) do { kprintf(MSG); while(1); } while (0);
+
+TAILQ_HEAD(vm_seglist, vm_phys_seg);
+static struct vm_seglist seglist;
+
+void vm_phys_print_free_blocks() {
+  vm_phys_seg_t *seg_it;
+  TAILQ_FOREACH(seg_it, &seglist, segq) {
+    for (int i = 0; i < VM_NFREEORDER; i++) {
+      kprintf("order %d:\n", i);
+      vm_page_t *pg_it;
+      TAILQ_FOREACH(pg_it, &seg_it->free_queues[i], freeq)
+      kprintf("(%p,%p) ", pg_it->phys_addr,
```

Comment:
I suggest creating helper macros:
* `pg_start(pg) ((pg)->phys_addr))`
* `pg_end(pg) ((pg)->phys_addr + (1 << (pg)->order))`

... and use them throughout your whole code. This will increase readability.